tsdb/head.go

```diff
@@ -53,13 +53,23 @@ var (
 	// ErrAppenderClosed is returned if an appender has already be successfully
 	// rolled back or committed.
 	ErrAppenderClosed = errors.New("appender closed")
+
+	WalReplayStatus walReplayStatus
```

Comment:
The method you're referring to relies on this data structure:

https://github.com/prometheus/prometheus/blob/bbc096cace886a5ca53cde72f56a36181e98d0b7/cmd/prometheus/main.go#L1056-L1062

Which is only populated with the `Head` once the instance is ready, and therefore only allows us to serve WAL replay status after it's already over. 

The current approach I'm leaning towards is having a variable in `head.go` like before, but having `api.go` call a function in `head.go` that reads it (after obtaining the lock), instead of having `api.go` read it directly like it currently does. This is the same way the variable containing the `readyStorage` value is handled.

Thoughts?

Reply:
Ah, that makes things more complicated. Definitely we should solve this without globals though.

I wonder if we could restructure the whole readiness handling such that `readyStorage` holds the TSDB instance from the beginning (so `db` would never be `nil`), but instead its wrapper methods would determine readiness by calling the `WALReplayStatus()` method on the TSDB. Then I'd also extend the `TSDBAdminStats` interface at https://github.com/prometheus/prometheus/blob/9549a15c6ffd1e997ee15f6262e5a215a64d5661/web/api/v1/api.go#L150-L157 with a new `WALReplayStatus()` method and implement that in `readyStorage` as well. Just that that method would not have the usual readiness handling that `Stats()` etc. have, since it's already supposed to work before the TSDB is ready. Then you could use that method from the web API.

What do you think?

Reply:
I think that this would be too much. I think we handle this differently without going to the global or changing the interfaces. This is very prometheus-tsdb specific.

Reply:
Oh sorry @roidelapluie I hadn't refreshed my page in a while before I submitted that comment.

I don't think that it would be too difficult, only requiring changes to a few functions/structers (`tsdb.Open`, `tsdb.open`, `*readyStorage.get`, `main.openDBWithMetrics`, `api.TSDBAdminStats` is all I can think of) and it shouldn't affect anything else beyond   `main.go`.

That being said, I would definitely be open to exploring other approaches.

Reply:
Thanks for the exploration & explanation! Yeah, generally, splitting the TSDB creation and initialization into two externally visible steps could be a decent way to go. Another option would be to pass a stats object into the `tsdb.Open()` call that is updated by the DB initialization, but can be read simultaneously by the caller.

Since we definitely need to touch the TSDB code in one way or another, I would just like to make sure that the TSDB maintainers @codesome and @bwplotka are aware of this. Maybe they also have a specific opinion / advice on how to most cleanly expose initialization progress.

Reply:
Ack, looking on this part.

Reply:
I checked and ideally we get rid of the global.

@juliusv idea sounds quite accessible.

>  Another option would be to pass a stats object into the tsdb.Open() call that is updated by the DB initialization, but can be read simultaneously by the caller.

Otherwise, I would pass an additional function/interface to `API` that will have a direct connection to WAL/Head. Head can have `WalReplayStatus()` and we directly pass this method during initialization in Prometheus cmd/prometheus/main.go - no need to go through readiness storage 🤗